platform/android/src/snapshotter/map_snapshotter.cpp

```diff
@@ -143,6 +155,134 @@ void MapSnapshotter::deactivateFilesource(JNIEnv& env) {
+                      std::string("Invalid index").c_str());
+    }
+    // Insert it below the current at that index
+    try {
+        layer->addToStyle(snapshotter->getStyle(), layers.at(index)->getID());
+    } catch (const std::runtime_error& error) {
+        jni::ThrowNew(
+            env, jni::FindClass(env, "com/mapbox/mapboxsdk/style/layers/CannotAddLayerException"), error.what());
+    }
+}
+
+void MapSnapshotter::addLayerBelow(JNIEnv& env, jlong nativeLayerPtr, const jni::String& below) {
+    assert(nativeLayerPtr != 0);
+
+    auto* layer = reinterpret_cast<Layer*>(nativeLayerPtr);
+    try {
+        layer->addToStyle(
+            snapshotter->getStyle(),
+            below ? mbgl::optional<std::string>(jni::Make<std::string>(env, below)) : mbgl::optional<std::string>());
+    } catch (const std::runtime_error& error) {
+        jni::ThrowNew(
+            env, jni::FindClass(env, "com/mapbox/mapboxsdk/style/layers/CannotAddLayerException"), error.what());
+    }
+}
+
+void MapSnapshotter::addLayerAbove(JNIEnv& env, jlong nativeLayerPtr, const jni::String& above) {
+    assert(nativeLayerPtr != 0);
+    auto* layer = reinterpret_cast<Layer*>(nativeLayerPtr);
+
+    // Find the sibling
+    auto layers = snapshotter->getStyle().getLayers();
+    auto siblingId = jni::Make<std::string>(env, above);
+
+    size_t index = 0;
+    for (auto l : layers) {
+        if (l->getID() == siblingId) {
+            break;
+        }
+        index++;
+    }
+
+    // Check if we found a sibling to place before
+    mbgl::optional<std::string> before;
+    if (index + 1 > layers.size()) {
+        // Not found
+        jni::ThrowNew(env,
+                      jni::FindClass(env, "com/mapbox/mapboxsdk/style/layers/CannotAddLayerException"),
+                      std::string("Could not find layer: ").append(siblingId).c_str());
+    } else if (index + 1 < layers.size()) {
+        // Place before the sibling
+        before = {layers.at(index + 1)->getID()};
+    }
+
+    // Add the layer
+    try {
+        layer->addToStyle(snapshotter->getStyle(), before);
+    } catch (const std::runtime_error& error) {
+        jni::ThrowNew(
+            env, jni::FindClass(env, "com/mapbox/mapboxsdk/style/layers/CannotAddLayerException"), error.what());
+    }
+}
+
+void MapSnapshotter::addSource(JNIEnv& env, const jni::Object<Source>& obj, jlong sourcePtr) {
+    assert(sourcePtr != 0);
+
+    auto* source = reinterpret_cast<Source*>(sourcePtr);
+    try {
+        source->addToStyle(env, obj, snapshotter->getStyle());
+    } catch (const std::runtime_error& error) {
+        jni::ThrowNew(
+            env, jni::FindClass(env, "com/mapbox/mapboxsdk/style/sources/CannotAddSourceException"), error.what());
+    }
+}
+
+void MapSnapshotter::addImages(JNIEnv& env, const jni::Array<jni::Object<mbgl::android::Image>>& jimages) {
+    jni::NullCheck(env, &jimages);
+    std::size_t len = jimages.Length(env);
+
+    for (std::size_t i = 0; i < len; i++) {
```

Comment:
++i

Reply:
It seems the same. Any reason why need to change to `++i`?

Reply:
@Chaoba https://stackoverflow.com/questions/24901/is-there-a-performance-difference-between-i-and-i-in-c